tests/testthat/test-code_dependency.R

```diff
@@ -0,0 +1,369 @@
+    )
+  }
+)
+
+testthat::test_that(
+  "get_code returns the same class when names is specified and when not",
+  {
+    q <- eval_code(new_qenv(), "a <- 1")
+    testthat::expect_identical(
+      get_code(q, names = "a"),
+      get_code(q)
+    )
+  }
+)
+
+testthat::test_that(
+  "get_code returns single lines for code put in {} inside expressions",
+  {
+    q <- new_qenv()
+    q <- eval_code(q, expression({
+      a <- 1
+      b <- 2
+    }))
+
+    testthat::expect_identical(
+      get_code(q, names = "a"),
+      "a <- 1"
+    )
+  }
+)
+
+
+
+# functions -------------------------------------------------------------------------------------------------------
+
+testthat::test_that("get_code ignores occurrence in function definition", {
+  q <- new_qenv()
+  q <- eval_code(q, "b <- 2")
+  q <- eval_code(q, "foo <- function(b) { b <- b + 2 }")
+
+  testthat::expect_identical(
+    get_code(q, names = "foo"),
+    c("foo <- function(b) {", "  b <- b + 2", "}")
+  )
+})
+
+
+testthat::test_that("get_code ignores effect of the object which occurs in a function definition", {
+  q <- new_qenv()
+  q <- eval_code(q, "b <- 2")
+  q <- eval_code(q, "foo <- function(b) { b <- b + 2 }")
+
+  testthat::expect_identical(
+    get_code(q, names = "b"),
+    c("b <- 2")
+  )
+})
+
+testthat::test_that("get_code detects occurrence of the function object", {
+  q <- new_qenv()
+  q <- eval_code(q, "a <- 1")
+  q <- eval_code(q, "b <- 2")
+  q <- eval_code(q, "foo <- function(b) { b <- b + 2 }")
+  q <- eval_code(q, "b <- foo(a)")
+
+  testthat::expect_identical(
+    get_code(q, names = "b"),
+    c("a <- 1", "b <- 2", "foo <- function(b) {", "  b <- b + 2", "}", "b <- foo(a)")
+  )
+})
+
+testthat::test_that(
+  "Can't detect occurrence of function definition when a formal is named the same as a function",
+  {
+    skip("This does not return foo definition YET!")
+    q <- new_qenv()
+    q <- eval_code(q, "x <- 1")
+    q <- eval_code(q, "foo <- function(foo = 1) 'text'")
+    q <- eval_code(q, "a <- foo(x)")
+
+    testthat::expect_identical(
+      get_code(q, names = "a"),
+      c("x <- 1", "foo <- function(foo = 1) 'text'", "a <- foo(x)")
+    )
+  }
+)
+
+# $ ---------------------------------------------------------------------------------------------------------------
+
+
+testthat::test_that("get_code understands $ usage and do not treat rhs of $ as objects (only lhs)", {
+  q <- new_qenv()
+  q <- eval_code(q, "x <- data.frame(a = 1:3)")
+  q <- eval_code(q, "a <- data.frame(y = 1:3)")
+  q <- eval_code(q, "a$x <- a$y")
+  q <- eval_code(q, "a$x <- a$x + 2")
+  q <- eval_code(q, "a$x <- x$a")
+
+  testthat::expect_identical(
+    get_code(q, names = "x"),
+    c("x <- data.frame(a = 1:3)")
+  )
+  testthat::expect_identical(
+    get_code(q, names = "a"),
+    c("x <- data.frame(a = 1:3)", "a <- data.frame(y = 1:3)", "a$x <- a$y", "a$x <- a$x + 2", "a$x <- x$a")
+  )
+})
```

Comment:
I'm pretty surprised that this works. Feel free to add another test


```suggestion
})

testthat::test_that("get_code detects cooccurrence properly even if all objects are on rhs", {
  q <- new_qenv()
  q <- eval_code(q, "a <- 1")
  q <- eval_code(q, "b <- list(c = 2)")
  q <- eval_code(q, "b[[a]] <- 3")

  testthat::expect_identical(
    get_code(q, names = "b"),
    c("a <- 1", "b <- list(c = 2)", "b[[a]] <- 3")
  )
})
```

Reply:
yeah, it looks for SYMBOLS no matter on which side are they. will include such example in the code. thanks